Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -1178,130 +1225,122 @@ u32 GCMemcard::ReadAnimRGBA8(u8 index, u32* buffer, u8* delays) const
+  switch (bnr_format)
   {
   case 1:
-    animData += 96 * 32 + 2 * 256;  // image+palette
+    image_offset += 96 * 32 + 2 * 256;  // image+palette
     break;
   case 2:
-    animData += 96 * 32 * 2;
+    image_offset += 96 * 32 * 2;
     break;
   }
 
-  int fmts[8];
-  u8* data[8];
-  int frames = 0;
-
-  for (int i = 0; i < 8; i++)
+  // decode frame count and formats
+  const int max_frame_count = 8;
+  const u32 pixels_per_frame = 32 * 32;
+  const u32 palette_indices = 256;
+  u32 data_length = 0;
+  int frame_count = 0;
+  std::array<int, max_frame_count> frame_formats;
+  std::array<u8, max_frame_count> frame_delays;
+  std::array<u32, max_frame_count> frame_offsets;
+  bool has_shared_palette = false;
+  for (int i = 0; i < max_frame_count; ++i)
   {
-    fmts[i] = (formats >> (2 * i)) & 3;
-    delays[i] = ((fdelays >> (2 * i)) & 3);
-    data[i] = animData;
+    frame_formats[i] = (formats >> (2 * i)) & 3;
+    frame_delays[i] = ((fdelays >> (2 * i)) & 3);
+    delays[i] = frame_delays[i];
+    frame_offsets[i] = data_length;
 
-    if (!delays[i])
+    if (frame_delays[i] == 0)
     {
       // First icon_speed = 0 indicates there aren't any more icons
       break;
     }
     // If speed is set there is an icon (it can be a "blank frame")
-    frames++;
-    if (fmts[i] != 0)
+    ++frame_count;
+    if (frame_formats[i] != 0)
     {
-      switch (fmts[i])
+      switch (frame_formats[i])
       {
       case CI8SHARED:  // CI8 with shared palette
-        animData += 32 * 32;
+        data_length += pixels_per_frame;
+        has_shared_palette = true;
         break;
       case RGB5A3:  // RGB5A3
-        animData += 32 * 32 * 2;
+        data_length += pixels_per_frame * 2;
         break;
       case CI8:  // CI8 with own palette
-        animData += 32 * 32 + 2 * 256;
+        data_length += pixels_per_frame + 2 * palette_indices;
         break;
       }
     }
   }
 
-  const u16* sharedPal = reinterpret_cast<u16*>(animData);
+  const u32 shared_palette_offset = data_length;
+  if (has_shared_palette)
+    data_length += 2 * palette_indices;
 
-  for (int i = 0; i < 8; i++)
+  auto save_data_bytes = GetSaveDataBytes(index, image_offset, data_length);
+  if (!save_data_bytes || save_data_bytes->size() != data_length)
+    return 0;
+
+  std::array<u16, palette_indices> shared_palette;
+  if (has_shared_palette)
   {
-    if (!delays[i])
-    {
-      // First icon_speed = 0 indicates there aren't any more icons
-      break;
-    }
-    if (fmts[i] != 0)
+    std::memcpy(shared_palette.data(), save_data_bytes->data() + shared_palette_offset,
+                2 * palette_indices);
+  }
+
+  for (int i = 0; i < frame_count; ++i)
+  {
+    // If speed is set but there's no actual icon show the next valid icon.
+    // This is used to reduce animation speed in Pikmin and Luigi's Mansion for example.
+    // TODO: Does this really show the *next* icon instead of leaving the previous on-screen?
```

Comment:
both games have a setup of [frame][3 x blank][frame][2 x blank][frame] ( 0x4041) , ping pong style , every frame lasting 12 frames (its delay settings is 0xFFFF)
for luigi's mansion this is luigi turning his head every ~2s ( in the GC's BIOS )
it has 2 frames with his head to the left (frame 1 & 5 ) and 1 to the right ( frame 8 )

if a blank is the previous frame, in 1 ping pong it would show the left 12 frames ( * 12 frames per fps, 2.4s)  and right only 3 ( 0.6s ).
showing next frame would result in left 9 ( 1,8s ) and right 6 ( 1,2s ).
there is still something im missing as 1.2s right is nowhere near the 2s im seeing in BIOS, but with the other PR's together it seems to work out if its next frame